### PR TITLE
Master sale expense purchase price v2 sgv

### PR DESCRIPTION
**Pull request by sofiagvaladze**
Set purchase_price on sale.order.line for expenses

### CODE REVIEW

File: addons/sale_expense_margin/models/sale_order_line.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    expense_id = fields.Many2one('hr.expense', string='Expense')
+
+    @api.depends('is_expense')
+    def _compute_purchase_price(self):
+        date_today = fields.Date.context_today(self.env.user)
+        expense_lines = self.filtered(lambda sol: sol.expense_id)
```
Comment (Feyensv) [inline addons/sale_expense_margin/models/sale_order_line.py:14]:
```suggestion
        expense_lines = self.filtered('expense_id')
```

File: addons/sale_expense_margin/models/sale_order_line.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        for line in expense_lines:
+            if line.expense_id.product_has_cost:
+                product_cost = line.expense_id.untaxed_amount / line.expense_id.quantity
+            else:
+                product_cost = line.expense_id.untaxed_amount
+
+            from_currency = line.expense_id.currency_id or self.env.company.currency_id
```
Comment (Feyensv) [inline addons/sale_expense_margin/models/sale_order_line.py:21]:
```suggestion
            from_currency = line.expense_id.currency_id
```
The currency_id is required on the expenses, is the fallback really necessary ?

File: addons/sale_expense_margin/models/sale_order_line.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+from odoo import api, models, fields
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    expense_id = fields.Many2one('hr.expense', string='Expense')
+
+    @api.depends('is_expense')
+    def _compute_purchase_price(self):
+        date_today = fields.Date.context_today(self.env.user)
```
Comment (Feyensv) [inline addons/sale_expense_margin/models/sale_order_line.py:13]:
```suggestion
        date_today = fields.Date.context_today(self)
```
`context_today` only uses the environment of the given record, giving the user doesn't change anything

File: addons/sale_expense_margin/models/account_move.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models
+
+
+class AccountMoveLine(models.Model):
+    _inherit = 'account.move.line'
+
+    def _sale_prepare_sale_line_values(self, order, price):
+        self.ensure_one()
```
Comment (Feyensv) [inline addons/sale_expense_margin/models/account_move.py:11]: the super call already checks the ensure_one, no need to call it here :)

**Review by Feyensv — COMMENTED** [review]
Minor questions / linting remarks :)

Could you add some tests to cover this new code/logic please ? :pray:

**Comment by sofiagvaladze** [comment]
Thank you for all the suggestions @Feyensv :raised_hands:

**Comment by Feyensv** [comment]
Thanks for the test :+1:

**Comment by kba-odoo** [comment]
robodoo r+

**Comment by kba-odoo** [comment]
robodoo rebase-ff

**Pull request closed without merging**